### PR TITLE
feat(build-cli): Add exec command

### PR DESCRIPTION
**Pull request by tylerbutler**
This PR adds an `exec` command to build-cli, which can be used to execute shell commands in the context of packages or release groups in the repo.

The following example runs `pnpm -r remove @rushstack/eslint-config` on all independent packages and all release group root packages:

```shell
flub exec --all --releaseGroupRoots "pnpm -r remove @rushstack/eslint-config"
```

I think that for this command to be useful, one needs to be able to scope the packages that it runs on to the following groups:

1. All independent packages in the repo - use the `--packages` flag.
1. All packages in the repo using a particular scope - use the `--scope` and `--skipScope` flags.
1. A single package in a directory - use the `--dir` flag.
1. All individual packages in a release group - use the `--releaseGroup` flag.
1. All release group root packages - use the `--releaseGroup` and `--releaseGroupRoots` flags together.
1. All release group root packages and independent packages - use the `--all` and `--releaseGroupRoots` flags together.
1. All release group packages and independent packages - use the `--all` flag.

I implemented this by adding two new flags to the BasePackageCommand class, `--all` and `--releaseGroupRoots`. Most of the code change was in the base class. This means that the `typetest` command can also benefit from the new flags. That said, releaseGroupRoots in particular doesn't make sense for typetests, so perhaps the new flags shouldn't be on the base class? Feedback welcome.

### CODE REVIEW

**Comment by CraigMacomber** [comment]
Making the new options work for everything (what you did) seems like the best choice. I think it would be fine to normalize type test information across all release packages after a tooling update for example.

**Comment by tylerbutler** [comment]
> Making the new options work for everything (what you did) seems like the best choice. I think it would be fine to normalize type test information across all release packages after a tooling update for example.

That's a great scenario that I had not considered. Cool!

File: build-tools/packages/build-cli/src/BasePackageCommand.ts [hunk 1]
```diff
@@ -19,26 +46,39 @@ export abstract class PackageCommand<
...
+		}),
 		dir: Flags.directory({
 			char: "d",
 			description:
-				"Run on the package in this directory. Cannot be used with --releaseGroup or --packages.",
-			exclusive: ["packages", "releaseGroup"],
+				"Run on the package in this directory. Cannot be used with --all, --packages, or --releaseGroup.",
+			exclusive: ["packages", "releaseGroup", "all"],
 		}),
 		packages: Flags.boolean({
```
Comment (CraigMacomber) [inline build-tools/packages/build-cli/src/BasePackageCommand.ts:61]: We should probably leave it for a separate PR if we change it, but I find "packages" a confusing option. Renaming this to "independent" or "separate" or "non-grouped" seems like it would be better.

Comment (tylerbutler) [reply build-tools/packages/build-cli/src/BasePackageCommand.ts:61]: Good point. I'll do that in a separate PR.

File: build-tools/packages/build-cli/src/BasePackageCommand.ts [hunk 2]
```diff
@@ -11,6 +11,33 @@ import path from "node:path";
...
+type PackageKind =
+	/**
+	 * Package is an independent package.
+	 */
+	| "independentPackage"
+
+	/**
+	 * Package is part of a release group
+	 */
+	| "releaseGroupPackage"
```
Comment (CraigMacomber) [inline build-tools/packages/build-cli/src/BasePackageCommand.ts:29]: This means release group, but not the root of it right? Someone could think that this refers to the package for a release group, which is the root. I think "releaseGroupChild" would be more clear.

Also since all of these are about packages, we should probably make all their names include "package" or none of them, for consistency.

File: build-tools/packages/build-cli/src/BasePackageCommand.ts [hunk 3]
```diff
@@ -11,6 +11,33 @@ import path from "node:path";
...
+	/**
+	 * Package is the root package of a release group.
+	 */
+	| "releaseGroupRoot"
+
+	/**
+	 * Package is being loaded from a directory. The package may be one of the other three kinds. This kind is only used
+	 * when running on a package diurectly using its directory.
+	 */
+	| "packageDir";
```
Comment (CraigMacomber) [inline build-tools/packages/build-cli/src/BasePackageCommand.ts:40]: avoiding abbreviations is preferable for readability and reduced ambiguity. calling this "fromPath" or "directory" might be more clear.

File: build-tools/packages/build-cli/src/BasePackageCommand.ts [hunk 4]
```diff
@@ -127,31 +173,70 @@ export abstract class PackageCommand<
 		}
 	}
 
+	/**
+	 * Runs processPackage for each package in a release group, exlcuding the root.
+	 */
+	private async processReleaseGroup(
```
Comment (CraigMacomber) [inline build-tools/packages/build-cli/src/BasePackageCommand.ts:179]: I think separating the concerns of finding the right packages, and running an operation on them would be cleaner. Having this return a list of packages, then having the user of this add it to the full set its going to process would accomplish this.

This approach would also make actions like detecting duplicates, or applying multiple different operations to a set of packages easier.

It would also allow parallelizing across multiple release groups while having a single scheduler to avoid having too many in progress tasks and mixed status reporting.

File: build-tools/packages/build-cli/src/BasePackageCommand.ts [hunk 5]
```diff
@@ -127,31 +173,70 @@ export abstract class PackageCommand<
...
+		return this.processPackages(
+			ctx.packagesInReleaseGroup(releaseGroup).map((p) => p.directory),
+			{ kind: "releaseGroupPackage" },
+		);
+	}
+
+	/**
+	 * Runs processPackage for each independent package in the repo.
+	 */
+	private async processIndependentPackages(): Promise<void> {
```
Comment (CraigMacomber) [inline build-tools/packages/build-cli/src/BasePackageCommand.ts:200]: Same as above: I think having this return the list of found packages would be better.

File: build-tools/packages/build-cli/src/BasePackageCommand.ts (hunk 4 above)
Comment (CraigMacomber) [reply build-tools/packages/build-cli/src/BasePackageCommand.ts:179]: Also making async methods have useful return values tends to help reduce the risk or unawaited promises or unexpected concurrency.

File: build-tools/packages/build-cli/src/BasePackageCommand.ts [hunk 6]
```diff
@@ -127,31 +161,105 @@ export abstract class PackageCommand<
...
+		rootPackageOnly: boolean,
+	): PackageDetails[] {
+		this.info(`Finding packages for release group: ${releaseGroup}`);
+		if (rootPackageOnly) {
+			const rg = ctx.repo.releaseGroups.get(releaseGroup);
+			assert(rg !== undefined);
+			return [{ directory: rg.repoPath, kind: "releaseGroupRootPackage" }];
+		}
+
+		return ctx.packagesInReleaseGroup(releaseGroup).map((p) => {
```
Comment (CraigMacomber) [inline build-tools/packages/build-cli/src/BasePackageCommand.ts:179]: Again, for another PR, but I think a decent amount of logic over in monoRepo.ts could be replaced by a call to [releaseGroups](https://www.npmjs.com/package/find-packages) to find these packages.

**Review by CraigMacomber — APPROVED** [review]
